src/common/Fixed.ml

```diff
@@ -0,0 +1,236 @@
+(** This module defines the signatures and `Make` functors for the 'fixed point'
+    (or two-level) type we use for our intermediate representations
+*)
+open Core_kernel
+
+(** The fixed-point of `Pattern.t` annotated with some meta-data *)
+module type S = sig
+  module Pattern : Pattern.S
+
+  type 'a t = {pattern: 'a t Pattern.t; meta: 'a}
+  [@@deriving compare, map, fold, hash, sexp]
+
+  val fix : 'a * 'a t Pattern.t -> 'a t
+  val unfix : 'a t -> 'a * 'a t Pattern.t
+  val pattern_of : 'a t -> 'a t Pattern.t
```

Comment:
Or this? Given we're exposing record types.

Reply:
As above; `pattern_of` is already used a lot in the refactor and I used `unfix` a lot in the previous full refactor.